compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -236,107 +227,332 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
+
+  // Boolen to store if the mainloop can be pipelined (coarse-grained
+  // scheduling) and the instructions can be interleaved (fine-grained
+  // scheduling).
+  bool isSchedulable = false;
+
+  //
+  // Methods
+  //
+
+  // Populates the dependent operations in ``dependentOps`` for the given a op
+  // recursively that are in the same block and not added to the backward slice
+  // of some other op.
+  void backwardSliceOfDependentOps(llvm::SetVector<Operation*>& dependentOps,
+                                   Operation* op, Block* block) {
+    if (!seenDepOps.insert(op)) return;
+    // Add the unseen op to the dependentOps and recurse on its operands.
+    dependentOps.insert(op);
+    for (Value operand : op->getOperands()) {
+      Operation* defOp = operand.getDefiningOp();
+      if (defOp && defOp->getBlock() == block)
+        backwardSliceOfDependentOps(dependentOps, defOp, block);
+    }
   }
-  if (!isa<nvgpu::MmaSyncOp>(use->getOwner())) return false;
-  return use->getOperandNumber() == index;
-}
 
-/// Return true if the op is used as operand A of an mma.sync op.
-static bool isAOperand(Operation* op) { return isMMAOperand(op, 0); }
-/// Return true if the op is used as operand B of an mma.sync op.
-static bool isBOperand(Operation* op) { return isMMAOperand(op, 1); }
+  // Backtrack from the MmaSyncOp operand (mlir::OpOperand) to its defining
+  // mlir::Operation to find the ldmatrix or ld.shared operations that load
+  // MmaSyncOp operands.
+  void backtrackToFindSmemLoad(Operation* op,
+                               llvm::SetVector<Operation*>& loadOperations,
+                               Block* block) {
+    if (!op) return;
 
-/// Return a pipelining schedule that gives good performance on Nvidia
-/// Ampere target.
-static void getNvidiaTensorCorePipeline(
-    scf::ForOp forOp, std::vector<std::pair<Operation*, unsigned>>& ops,
-    unsigned depth) {
-  bool loopCanBePipelined = false;
-  // TODO: Tune this and make it a more fine grain logic.
-  static constexpr int64_t numPrefetchSmemLoadPerOperand = 4;
-  SmallVector<Operation*> stageCopyToSharedMemory;
-  SmallVector<Operation*> stagePrefetch;
-  SmallVector<Operation*> stageCompute;
-  int64_t numLoadA = 0;
-  int64_t numLoadB = 0;
-  for (Operation& op : forOp.getBody()->getOperations()) {
-    // Pipeline the most inner for op that should be a flat region.
-    if (op.getNumRegions() > 0) {
-      loopCanBePipelined = false;
-      break;
+    if (isa<nvgpu::LdMatrixOp, memref::LoadOp>(op)) {
+      if (op->getBlock() == block) {
+        loadOperations.insert(op);
+      }
+      return;
     }
-    if (isa<gpu::BarrierOp, nvgpu::DeviceAsyncWaitOp>(op)) {
-      stagePrefetch.push_back(&op);
+
+    // Recurse upwards towards the definition until a Shared Memory load is
+    // found. Assumption here is that only single operand operations are
+    // leading up to a Shared Memory load.
+    Operation* defOp = op->getOperand(0).getDefiningOp();
```

Comment:
This is not safe as it assumes the op has an operand without know what it is.

Reply:
Check the checks in the new version :)